unit-tests/live/frames/test-fps-permutations.py

```diff
@@ -0,0 +1,304 @@
+    Creates callable functions for each sensor to be triggered when a new frame arrives
+    Used to count frames received for measuring fps
+    """
+    sensor_function_dict = {}
+    for sensor_key in sensor_profiles_dict:
+        def on_frame_received(frame, key=sensor_key):
+            global count_frames
+            if count_frames:
+                sensor_profiles_dict[key] += 1
+
+        sensor_function_dict[sensor_key] = on_frame_received
+    return sensor_function_dict
+
+
+def measure_fps(sensor_profiles_dict):
+    """
+    Given a dictionary of sensors and profiles to test, activate all sensors on the given profiles
+    and measure fps
+    Return a dictionary of sensors and the fps measured for them
+    """
+    global seconds_till_steady_state
+    global seconds_to_count_frames
+
+    global count_frames
+    count_frames = False
+
+    # initialize fps dict
+    sensor_fps_dict = {}
+    for key in sensor_profiles_dict:
+        sensor_fps_dict[key] = 0
+
+    # generate sensor-callable dictionary
+    funcs_dict = generate_functions(sensor_fps_dict)
+
+    for key in sensor_profiles_dict:
+        sensor = key
+        profiles = sensor_profiles_dict[key]
+        sensor.open(profiles)
+        sensor.start(funcs_dict[key])
+
+    # the core of the test - frames are counted during sleep when count_frames is on
+    time.sleep(seconds_till_steady_state)
+    count_frames = True  # Start counting frames
+    time.sleep(seconds_to_count_frames)
+    count_frames = False  # Stop counting
+
+    for key in sensor_profiles_dict:
+        sensor_fps_dict[key] /= len(sensor_profiles_dict[key])  # number of profiles on the sensor
+        sensor_fps_dict[key] /= seconds_to_count_frames
+        # now for each sensor we have the average fps received
+
+        sensor = key
+        sensor.stop()
+        sensor.close()
+
+    return sensor_fps_dict
+
+
+def get_test_details_str(sensor_profile_dict, expected_fps_dict):
+    s = ""
+    for sensor_key in sensor_profile_dict:
+        if len(sensor_profile_dict[sensor_key]) > 1:
+            s += f"Expected average fps for {sensor_key.name} is {expected_fps_dict[sensor_key]}:\n"
+            for profile in sensor_profile_dict[sensor_key]:
+                s += f"Profile {profile.stream_name()} expects {profile.fps()} fps on {get_resolution(profile)}\n"
+        else:
+            s += (f"Expected fps for sensor {sensor_key.name} on profile "
+                  f"{sensor_profile_dict[sensor_key][0].stream_name()} is {expected_fps_dict[sensor_key]}"
+                  f" on {get_resolution(sensor_profile_dict[sensor_key][0])}\n")
+    s = s.replace("on (0, 0)", "")  # remove no resolution for Motion Module profiles
+    s += "***************"
+    return s
+
+
+def perform_fps_test(sensor_profiles_arr, modes):
+
+    log.d(get_time_est_string(len(sensor_profiles_arr), modes))
+
+    for mode in modes:
+        perms = list(itertools.product([0, 1], repeat=len(sensor_profiles_arr)))
```

Comment:
Is `itertools` native python?
Please explain what does it do?

Reply:
Yes, it's native python.
It's generating all possible combinations of 1s and 0s in the length of the array, I'll add a comment explaining that